internalservice_agentmanager.go

```diff
@@ -0,0 +1,331 @@
+		HttpRoundTripper: httpTransport,
+		Endpoints:        srcHTTPAddrs,
+		UserAgent:        httpUserAgent,
+		Authenticator:    opts.Authenticator,
+	}, &StreamWatcherHttpOptions{
+		Logger: logger.Named("buckets-watcher"),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	m.bucketsWatcher = bucketsWatcher
+
+	topoWatcher, err := NewConfigWatcherHttp(
+		&ConfigWatcherHttpConfig{
+			HttpRoundTripper: httpTransport,
+			Endpoints:        srcHTTPAddrs,
+			UserAgent:        httpUserAgent,
+			Authenticator:    opts.Authenticator,
+		},
+		&ConfigWatcherHttpOptions{
+			Logger: logger.Named("topology-watcher"),
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	m.topologyCfgWatcher = topoWatcher
+
+	clusterAgent, err := m.makeAgentLocked(ctx, "")
+	if err != nil {
+		return nil, err
+	}
+
+	m.clusterAgent = clusterAgent
+
+	m.startWatchers()
+
+	return m, nil
+}
+
+func (m *InternalServiceAgentManager) startWatchers() {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	stoppedTopoSig := make(chan struct{})
+	stoppedBucketSig := make(chan struct{})
+	m.watchersCancel = func() {
+		// Cancel the context and wait for the configCh to be closed.
+		cancel()
+		<-stoppedBucketSig
+		<-stoppedTopoSig
+	}
+
+	bucketCh := m.bucketsWatcher.Watch(ctx, streamWatcherHttp_streamBuckets)
+
+	go func() {
+		for bucket := range bucketCh {
+			m.handleBuckets(bucket)
+		}
+		close(stoppedBucketSig)
+	}()
+
+	topoCh := m.topologyCfgWatcher.Watch(ctx)
+
+	go func() {
+		for config := range topoCh {
+			m.applyConfig(config)
+		}
+		close(stoppedTopoSig)
+	}()
+}
+
+func (m *InternalServiceAgentManager) applyConfig(config *ParsedConfig) {
+	m.topoLock.Lock()
+	defer m.topoLock.Unlock()
+
+	if !canUpdateConfig(config, m.state.latestConfig, m.opts.Logger) {
+		return
+	}
+
+	m.logger.Debug("Applying new config", zap.Int64("revId", config.RevID), zap.Int64("revEpoch", config.RevEpoch))
+
+	m.state.latestConfig = config
+
+	bootstrapHosts := m.state.latestConfig.AddressesGroupForNetworkType(m.networkType)
+
+	var mgmtEndpoints []string
+	tlsConfig := m.state.tlsConfig
+	if tlsConfig == nil {
+		for _, host := range bootstrapHosts.NonSSL.Mgmt {
+			mgmtEndpoints = append(mgmtEndpoints, "http://"+host)
+		}
+	} else {
+		for _, host := range bootstrapHosts.SSL.Mgmt {
+			mgmtEndpoints = append(mgmtEndpoints, "https://"+host)
+		}
+	}
+
+	m.topologyCfgWatcher.Reconfigure(&ConfigWatcherHttpConfig{
+		HttpRoundTripper: m.state.httpTransport,
+		Endpoints:        mgmtEndpoints,
+		UserAgent:        m.userAgent,
+		Authenticator:    m.state.authenticator,
+	})
+
+	m.bucketsWatcher.Reconfigure(&StreamWatcherHttpConfig{
+		HttpRoundTripper: m.state.httpTransport,
+		Endpoints:        mgmtEndpoints,
+		UserAgent:        m.userAgent,
+		Authenticator:    m.state.authenticator,
+	})
+}
+
+func (m *InternalServiceAgentManager) handleBuckets(buckets []bucketDescriptor) {
+	m.bucketsLock.Lock()
```

Comment:
We should probably file a bug to cover the fact that this implementation still has the same problem the original did where it ends up holding a 'global' lock when creating agents, making it so that you cant fetch the bucketA while bucketB is being connected.  This probably is trivially solvable by using a similar approach as we have with collections, with the slow map, and a fast lookup table.

Reply:
https://github.com/couchbase/gocbcorex/issues/159